modules/core/src/lib/layer.js

```diff
@@ -402,13 +404,35 @@ export default class Layer extends Component {
 
   calculateInstancePickingColors(attribute, {numInstances}) {
     const {value, size} = attribute;
-    // add 1 to index to seperate from no selection
-    for (let i = 0; i < numInstances; i++) {
+
+    if (value[0] === 1) {
+      // This can happen when data has changed, but the attribute value typed array
+      // has sufficient size and does not need to be re-allocated.
+      // This attribute is already populated, we do not have to recalculate it
+      return;
+    }
+
+    const cacheSize = pickingColorCache.length / size;
+
+    // Copy the last calculated picking color sequence into the attribute
+    value.set(
+      numInstances < cacheSize
+        ? pickingColorCache.subarray(0, numInstances * size)
+        : pickingColorCache
+    );
+
+    // If the attribute is larger than the cache, populate the missing chunk
+    for (let i = cacheSize; i < numInstances; i++) {
       const pickingColor = this.encodePickingColor(i);
       value[i * size + 0] = pickingColor[0];
       value[i * size + 1] = pickingColor[1];
       value[i * size + 2] = pickingColor[2];
     }
+
+    if (cacheSize < numInstances) {
+      // Save the largest picking color sequence
+      pickingColorCache = value;
```

Comment:
This seems like an error. Do you really want pickingColorCache to reference the array created internally in the attribute?